### ISSUE TITLE
Multi-stage builds

### ISSUE THREAD

**Issue by martinbaillie** [docker/for-mac#1751]
I have a simple multi-stage Dockerfile very similar to the example at: https://docs.docker.com/engine/userguide/eng-image/multistage-build/#name-your-build-stages

It builds perfectly on a Linux machine with: 
> Docker version v17.05.0-ce, (89658be)

But, using the exact same codebase and Dockerfile on:
> Docker for-mac edge v17.06.0-rc4-ce-mac15 (18357)

It fails at the `COPY --from=builder` directive with: `invalid from flag value builder: repository sha256 not found: does not exist or no pull access`

Has anyone else tried this? I'm unclear if this is a bug or a known limitation?

**Comment by Schnitzel** [comment]
@martinbaillie we have the same problem, but actually also on Linux v17.06.0. The issue comes when we try to build a multistage build with the `--pull` parameter.

Can you try if 
```
docker build --pull .
```
fails

and 
```
docker build .
```
works?

We might need to report that as a Docker general bug.

**Comment by thaJeztah** [comment]
ping @dnephin @tonistiigi @nishanttotla PTAL

**Comment by tonistiigi** [comment]
@martinbaillie @Schnitzel Can you post a `Dockerfile`. Are you accessing build stages by image ID?

**Comment by Schnitzel** [comment]
@tonistiigi 
yes, this one for example: https://github.com/amazeeio/node-example/blob/master/.amazeeio.Dockerfile

and here the logs once with and without `--pull`
https://gist.github.com/Schnitzel/393ca94c85394f69c02bf4659aa31143

**Comment by Schnitzel** [comment]
I think the bug is that it tries to parse `sha256:d1b3fd8913d1af618fa71e2daec36e2e6387d201eadffb443da7f1b427e32d25` as a Docker Repo URL and thinks that `sha256` is the host of the docker repo :)

**Comment by martinbaillie** [comment]
@Schnitzel you're absolutely right. I was using `--pull`. 

Can confirm that when run without `--pull` all is well 👍

**Comment by soar** [comment]
Still not working with fresh Docker:

```
Client:
 Version:      17.06.0-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:23:31 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.0-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:19:04 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Comment by thaJeztah** [comment]
The fix for this will be included in the 17.06.1 patch release that's being prepared

**Comment by sethpollack** [comment]
@thaJeztah Did the fix make it into `v17.06.1-ce-rc1`?

**Comment by y3ti** [comment]
@sethpollack I have tested the issue with `17.06.1-ce-rc1`. Still doesn't work :(

```
$ docker version
Client:
 Version:      17.06.1-ce-rc1
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   77b4dce
 Built:        Fri Jul 14 07:38:15 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      17.06.1-ce-rc1
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   77b4dce
 Built:        Fri Jul 14 07:33:35 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Comment by thaJeztah** [comment]
It's not in rc1 yet

**Comment by SwenChan** [comment]
My mac use this file
```
FROM node:7-alpine as builder
WORKDIR /from
RUN echo ing && \
        echo test > text.txt && \
        echo end `pwd`

FROM node:7-alpine
WORKDIR /to
RUN ls
COPY --from=builder /from/text.txt .
RUN ls && pwd
```
and it just didn't work. The second 'FROM' just not run and the output is
```
Sending build context to Docker daemon  4.096kB
Step 1/8 : FROM node:7-alpine as builder
 ---> 4b72b56791f9
Step 2/8 : WORKDIR /from
 ---> d71f7b6335d6
Removing intermediate container 37fb57151e9f
Step 3/8 : RUN echo ing && 	echo test > text.txt && 	echo end `pwd`
 ---> Running in 4344ac929101
ing
end /from
```
my docker engine is:
```
Version 17.06.0-ce-mac19 (18663)
Channel: stable
c98c1c25e0
```
Can anyone help me? Multi-stage is a really awesome feature for me and I have been try so many time and it just not work.

**Comment by SwenChan** [comment]
sorry, just have some problem in my docker, I restart it and I build it success....

**Comment by iv597** [comment]
I ran into this problem, but took the update I had been putting off for ~a week now, which resolved it for me.

Multi-stage builds with `--pull` are working for me on:
```
Docker for Mac: version: 17.06.1-ce-mac24 (4cdec4294a50b2233146b09469b49937dabdebdd)
macOS: version 10.12.6 (build: 16G29)
```

**Comment by shatil** [comment]
After updating, multi-stage builds with `--pull` work great!

```
$ docker version
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:40:09 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:45:38 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

Previously, on this version, it didn't work:

```
$ docker version
Client:
 Version:      17.06.0-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:31:53 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      17.06.0-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:51:55 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Comment by thaJeztah** [comment]
Thanks for confirming, this should indeed be resolved by https://github.com/moby/moby/pull/33735, which is included in docker 17.06.1 and up.

Closing this issue as resolved 👍
